docs: storybook upgrade & renovation

KaiVandivier opened the pull request:
This is a project to take advantage of new Storybook features to renovate the UI library documentation and make it a more useful tool for readers.

Relates to [this proposal](https://github.com/dhis2/notes/discussions/197) at the 'notes' repository

KaiVandivier commented:
#### Notes from developer review on 18 Feb
* Regarding the existing API docs
    * It's fast to load & deep linkable - some like having it available
    * Some people believe it to be user unfriendly, though
    * Storybook _does_ provide API info with args tables
    * Maybe API docs should be replaced by Storybook
    * Maybe API docs can use react docgen to generate a static site so that it uses the same component and prop documentation as storybook, and those comments don't need to be duplicated
* Arg types and colocation with shared prop types: currently fine how they are; ideally they will be 'shaken out' from the UI bundle when tree shaking is working
* Prop types
    * Okay to import from react's `prop-types` library
    * Note: I've switched out the `prop-types` for any component with user-facing stories -- components without stories or with only e2e stories haven't been changed. Should we switch them out for all the components?
    * There is some consideration for discontinuing re-exporting prop-types from the `@dhis2/prop-types` library
    * Maybe use a [react docgen handler](https://github.com/benjroy/react-docgen-imported-proptype-handler)
* `iframe` things (note that everything works fine in the Canvas tab)
    * 'Single representative example' approved 👍 
    * Maybe allow full-screen things to be full-screen, but open them in response to a button click, a la [Material UI](https://material-ui.com/components/modal/)
    * Is the dropdown placement for Selects and Poppers a bug with the component, or a quirk of the Storybook Docs Page? @HendrikThePendric @Mohammer5 


#### Other notes/questions about this content
* Importing markdown in MDX files works, so it's easy to mirror the `.md` files for the docsite in `.stories.mdx` files for the storybook. Maybe this could be done better with a script at build time that makes a `.stories.mdx` file for each markdown file (Todo)
* Should we move forward with making the storybook the main UI documentation source? @varl @amcgee 
    * This would affect JSDoc/component descriptions content - as mentioned above, we want to have only one of the two
    * This would also affect the need to make `.stories.mdx` files from `.md` files for documentation pages
    * If so, it would make sense to point `ui.dhis2.nu` straight to the storybook
* I made an attempt at theming and adding images and a favicon to the appearance, and I'm open to feedback @cooper-joe 
    * The theme definitions can be found in `/.storybook/theme.js` including comments about what they affect on the page - a few values are set. If it looks good, I'll clean up that file
    * [Theming storybook](https://storybook.js.org/docs/react/configure/theming) docs
    * A favicon is added using `/.storybook/manager-head.html` and some files in `/.storybook/static/`. Should the whole [favicon package](https://github.com/dhis2/dhis2-identity/tree/master/web/favicon_package) be added?
    * Static files are served for storybook by adding the `-s .storybook/static` option in storybook scripts ([docs about assets](https://storybook.js.org/docs/react/configure/images-and-assets))


#### Todo
* Add a page to instruct users how to best understand the storybook content
* Add pages describing UI constants
* It would be nice to show relevant pages from the Design System in the storybook for reference, but we want to avoid having those files in more than one place. Therefore, a way to fetch the pages and convert them into `.stories.mdx` at build time could accomplish this
* Make sure annotations, descriptions, and code samples are clear for all components

cooper-joe reviewed (changes requested):
Great stuff @KaiVandivier, thanks for putting this all together.

Some suggestions:

**1. Page titles**
I think it might be useful to prepend the page titles with `ui –`, as the DHIS2 logo favicon is used in other situations too, so having the `ui –` prefix would further clarify what DHIS2 domain this is.
In that case, page titles would look like: `ui – actions/button`, `ui – data display/stacked table`.

**2. Change "Stories" page heading**
Most component pages contain the heading "Stories". This is an internal storybook-specific term and doesn't have any meaning to the components. If this is possible to change, I'd suggest "Variants" instead, which more accurately describes the content.

**3. Adjust section titles**
I think the sidebar sections could be more descriptive. I suggest:
- Adding a heading for the "...for readers" and "...for maintainers" pages, with the title "About this documentation"
- Change the "Tips for readers" page title to just "For readers"
- Change the "Tips for maintainers" page title to just "For maintainers"
- Change 'UI docs" heading to "Using UI"

**4. Custom styles**
Am I correct in thinking the only styles we can adjust are those exposed in `theme.js`? There are some other changes I'd like to make (document link color contrast is not accessible, sidebar headers are difficult to read), but there doesn't seem to be any options available. Any workaround for this?

KaiVandivier commented:
Thanks for the great suggestions Joe!

1. Sounds good - I'm looking into how to change the page titles
2. Agreed - I tried to avoid storybook-specific vocabulary in the page for readers too. Would 'Examples' also work for that section heading?
3. 👍 
4. It turns out pretty much everything on the page can be configured style-wise: [CSS escape hatches](https://storybook.js.org/docs/react/configure/theming#css-escape-hatches)

Mohammer5 commented:
> Is the dropdown placement for Selects and Poppers a bug with the component, or a quirk of the Storybook Docs Page? @HendrikThePendric @Mohammer5

Sorry, I didn't see this before.. Could you elaborate what what is going wrong? I just took a brief look at the first single select story's docs tab and it seems to work fine in the iframe?

KaiVandivier commented:
@Mohammer5 In the SingleSelect, iframe works around the issue; try looking at SingleSelectField demos on the docs page (which don't use an iframe) - the first demo will usually work, but clicking on one of the later demos opens up the dropdown somewhere else on the page, and you will have to scroll up to see it

cooper-joe commented:
>     2. Agreed - I tried to avoid storybook-specific vocabulary in the page for readers too. Would 'Examples' also work for that section heading?

Yeah, I think `Examples` is probably a good catch-all, it works even when the examples are not strictly 'variants'.
 
>     4. It turns out pretty much everything on the page can be configured style-wise: [CSS escape hatches](https://storybook.js.org/docs/react/configure/theming#css-escape-hatches)

Ok, interesting! I think it's best to keep the scope of this PR as narrow as possible, so I can address the visual style changes I have in a future PR. Perhaps you could include the escape hatch functionality in this PR with a blank stylesheet? Then I can add to it in future.

KaiVandivier commented:
> Perhaps you could include the escape hatch functionality in this PR with a blank stylesheet? Then I can add to it in future.

Will do! 👍

KaiVandivier commented on .storybook/theme.js:
```diff
@@ -0,0 +1,93 @@
+import { create } from '@storybook/theming/create'
+import { colors } from '@dhis2/ui-constants'
+
+const DHIS2_ACCENT = "#43cbcb"
```
I tried `teal800` and it looked very dark... I tried `teal600` and I think it looks a little bit better, and I verified its contrast passed WCAG-AAA for bold text (which is how it's used in Storybook) and WCAG-AA for regular text - what do you think?

cooper-joe replied:
Yeah sure, that works if it's only on bold text 👍

KaiVandivier commented:
Updates:
1. Changed 'meta documentation' page titles and section names
2. Renamed 'Stories' heading to 'Examples' in docs pages
3. Added files for styling the UI with CSS escape hatches: `./storybook/static/main.css` linked by `manager-head.html`, and `./storybook/static/docs.css` linked by `preview-head.html` (...now that I've written that out, I think those I could have named css files better)
4. Document titles - [not easy to change](https://github.com/storybookjs/storybook/issues/6339), surprisingly - I'm still looking into it

KaiVandivier commented:
The document title might be solved by [writing an addon](https://storybook.js.org/docs/react/addons/writing-addons), but that might be best done in another pull request 🤔

varl reviewed (approved):
On the whole, I think this is an improvement that is worthy to go in. Parts that need to change can be dealt with in smaller PRs.

cooper-joe reviewed (approved):
Thanks for making the requested changes. This looks good! I'll be making some separate PR's to deal with minor styling and aesthetics that I noted.

amcgee commented:
🎉  great stuff @KaiVandivier !
